### PR TITLE
Scrapper, Olga Pershina - 21FPL1

### PR DESCRIPTION
**Pull request by OlyaP12**

### CODE REVIEW

**Review by marina-kaz — COMMENTED** [review]
Welcome to the CTLR course! 
Reviews will appear every Wednesday. 
Make sure to publish relevant changes by that day of the week. 
Good luck!

**Review by marina-kaz — COMMENTED** [review]
Nothing to review as of today

File: lab_5_scrapper/scrapper.py [hunk 1]
```diff
@@ -13,62 +52,93 @@ def __init__(self, path_to_config: Path) -> None:
...
-        pass
+        if not self.config_dto.seed_urls or not isinstance(self.config_dto.seed_urls, list) \
+                or not all(isinstance(url, str) for url in self.config_dto.seed_urls):
+            raise IncorrectSeedURLError
+
+        if not isinstance(self.config_dto.total_articles_to_find_and_parse, int) \
+                or not 1 < self.config_dto.total_articles_to_find_and_parse < 150:
+            raise IncorrectNumberOfArticlesError
+
+        if not self.config_dto.seed_urls and not isinstance(self.config_dto.headers, dict) \
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py]: ` if not self.config_dto.seed_urls`: it looks like you've already checked this condition in the 79th line

File: lab_5_scrapper/scrapper.py [hunk 2]
```diff
@@ -13,62 +52,93 @@ def __init__(self, path_to_config: Path) -> None:
...
+                and not all(isinstance(key, str) and isinstance(value, str)
+                            for key, value in self.config_dto.headers.items()):
+            raise IncorrectHeadersError
+
+        if not isinstance(self.config_dto.encoding, str):
+            raise IncorrectEncodingError
+
+        if not isinstance(self.config_dto.timeout, int) or self.config_dto.timeout < 1 or self.config_dto.timeout > 60:
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py]: consider using constants from `core_utils.constants` for the upper and lower limits

File: lab_5_scrapper/scrapper.py [hunk 3]
```diff
@@ -13,62 +52,93 @@ def __init__(self, path_to_config: Path) -> None:
...
         Ensure configuration parameters
         are not corrupt
         """
-        pass
+        if not self.config_dto.seed_urls or not isinstance(self.config_dto.seed_urls, list) \
+                or not all(isinstance(url, str) for url in self.config_dto.seed_urls):
+            raise IncorrectSeedURLError
+
+        if not isinstance(self.config_dto.total_articles_to_find_and_parse, int) \
+                or not 1 < self.config_dto.total_articles_to_find_and_parse < 150:
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py]: consider using constants from `core_utils.constants` for the upper and lower limits

File: lab_5_scrapper/scrapper.py [hunk 4]
```diff
@@ -13,62 +52,93 @@ def __init__(self, path_to_config: Path) -> None:
...
 
     def _validate_config_content(self) -> None:
         """
         Ensure configuration parameters
         are not corrupt
         """
-        pass
+        if not self.config_dto.seed_urls or not isinstance(self.config_dto.seed_urls, list) \
+                or not all(isinstance(url, str) for url in self.config_dto.seed_urls):
+            raise IncorrectSeedURLError
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py:101]: what if URLs provided are strings that do not match the URL pattern?

**Review by marina-kaz — COMMENTED** [review]
That's a good start! Keep going

File: lab_5_scrapper/scrapper.py [hunk 5]
```diff
@@ -13,70 +57,116 @@ def __init__(self, path_to_config: Path) -> None:
...
         """
-        pass
+        config_dto = self._extract_config_content()
+
+        if not isinstance(config_dto.seed_urls, list):
+            raise IncorrectSeedURLError
+
+        for url in config_dto.seed_urls:
+            if not isinstance(url, str) or not re.match(r'https?://.*/', url):
+                raise IncorrectSeedURLError
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py:105]: this exception should also be raised if there are no urls provided, it looks like you've completely removed this check

File: lab_5_scrapper/scrapper.py [hunk 6]
```diff
@@ -13,70 +57,116 @@ def __init__(self, path_to_config: Path) -> None:
...
 
     def get_headless_mode(self) -> bool:
         """
         Retrieve whether to use headless mode
         """
-        pass
+        return self._headless_mode
 
 
 def make_request(url: str, config: Config) -> requests.models.Response:
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py:173]: consider also specifying encoding of the response in this method

File: lab_5_scrapper/scrapper.py [hunk 7]
```diff
@@ -13,70 +57,116 @@ def __init__(self, path_to_config: Path) -> None:
...
     """
     Delivers a response from a request
     with given configuration
     """
-    pass
+    time.sleep(random.randint(const.TIMEOUT_LOWER_LIMIT, const.TIMEOUT_UPPER_LIMIT))
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py]: do you really need to wait up to 60 seconds to make another request?

File: lab_5_scrapper/scrapper.py [hunk 8]
```diff
@@ -90,25 +180,37 @@ def __init__(self, config: Config) -> None:
...
+        return href
 
     def find_articles(self) -> None:
         """
         Finds articles
         """
-        pass
+        for url in self._config.get_seed_urls():
+            page = make_request(url, config=self._config)
+            page.encoding = self._config.get_encoding()
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py]: move this specification to the `make_request`

File: lab_5_scrapper/scrapper.py [hunk 9]
```diff
@@ -120,45 +222,83 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        self._config = config
+        self.article = Article(full_url, article_id)
 
     def _fill_article_with_text(self, article_soup: BeautifulSoup) -> None:
         """
         Finds text of article
         """
-        pass
+        content_bs = article_soup.find('div', {'class': 'news-detail__text'})
+        if content_bs:
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py:260]: is it possible that the text will not be collected? it is unacceptable

File: lab_5_scrapper/scrapper.py [hunk 10]
```diff
@@ -120,45 +222,83 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+                text += p.text.strip()
+            self.article.text = text
 
     def _fill_article_with_meta_information(self, article_soup: BeautifulSoup) -> None:
         """
         Finds meta information of article
         """
-        pass
+        title = article_soup.find('h1')
+        if title:
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py]: is it possible that the title field will not be filled? it is unacceptable

File: lab_5_scrapper/scrapper.py [hunk 11]
```diff
@@ -120,45 +222,83 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+            self.article.title = title.text.strip()
 
-    def unify_date_format(self, date_str: str) -> datetime.datetime:
+        span_author = article_soup.find('span', itemprop='name')
+        if span_author:
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py:276]: task description contains information concerning what to do if there is not author

File: lab_5_scrapper/scrapper.py [hunk 12]
```diff
@@ -120,45 +222,83 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+            self.article.author = span_author.text
+
+        div_date = article_soup.find_all('div', class_='news-detail__info-item')[1]
+        if div_date:
+            try:
+                self.article.date = self.unify_date_format(div_date.text.strip())
+            except ValueError:
+                pass
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py]: date field is also supposed to always be filled

File: lab_5_scrapper/scrapper.py [hunk 13]
```diff
@@ -120,45 +222,83 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
-        pass
+        return datetime.datetime.strptime(date_str, '%d.%m.%Y')
 
     def parse(self) -> Union[Article, bool, list]:
         """
         Parses each article
         """
-        pass
+        page = make_request(self.full_url, self._config)
+        page.encoding = self._config.get_encoding()
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py]: move this specification to the `make_request`

File: lab_5_scrapper/scrapper.py [hunk 14]
```diff
@@ -120,45 +222,83 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
         """
-        pass
+        return datetime.datetime.strptime(date_str, '%d.%m.%Y')
 
     def parse(self) -> Union[Article, bool, list]:
         """
         Parses each article
         """
-        pass
+        page = make_request(self.full_url, self._config)
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py:303]: what if request is unsuccessful?

File: lab_5_scrapper/scrapper.py [hunk 15]
```diff
@@ -120,45 +222,83 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
     """
     Entrypoint for scrapper module
     """
-    pass
+    config = Config(const.CRAWLER_CONFIG_PATH)
+    prepare_environment(const.ASSETS_PATH)
+    crawler = Crawler(config)
+    crawler.find_articles()
+    for id_, url in enumerate(crawler.get_search_urls()):
+        parser = HTMLParser(full_url=url, article_id=id_, config=config)
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py:403]: articles numbering must start from 1, not from 0

**Review by marina-kaz — COMMENTED** [review]
Good! Keep up the good work

File: lab_5_scrapper/scrapper.py (hunk 11 above)
Comment (OlyaP12) [reply lab_5_scrapper/scrapper.py:276]: done!

File: lab_5_scrapper/scrapper.py (hunk 14 above)
Comment (OlyaP12) [reply lab_5_scrapper/scrapper.py:303]: done!

File: lab_5_scrapper/scrapper.py (hunk 15 above)
Comment (OlyaP12) [reply lab_5_scrapper/scrapper.py:403]: corrected!

File: lab_5_scrapper/scrapper.py (hunk 6 above)
Comment (OlyaP12) [reply lab_5_scrapper/scrapper.py:173]: added!

File: lab_5_scrapper/scrapper.py (hunk 5 above)
Comment (OlyaP12) [reply lab_5_scrapper/scrapper.py:105]: added!

File: lab_5_scrapper/scrapper.py [hunk 16]
```diff
@@ -90,25 +193,45 @@ def __init__(self, config: Config) -> None:
...
     def find_articles(self) -> None:
         """
         Finds articles
         """
-        pass
+        start_url = self._config.get_seed_urls()[0]
+        url = start_url
+        count_of_page = 1
+        while len(self.urls) < self._config.get_num_articles():
+            page = make_request(url, config=self._config)
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py:217]: what if request is unsiccessfull?

**Review by marina-kaz — COMMENTED** [review]
Good job! A little more to fix

File: lab_5_scrapper/scrapper.py (hunk 16 above)
Comment (OlyaP12) [reply lab_5_scrapper/scrapper.py:217]: corrected

**Comment by vitazaytseva** [comment]
Well done! You can proceed to work on Lab 6)

**Pull request closed without merging**